File: fixtures/code/test-repo/app/models/base_model.rb
# Base model class for all application models
class BaseModel
  attr_reader :id, :created_at, :updated_at

  def initialize(attributes = {})
    @id = attributes[:id] || generate_id
    @created_at = attributes[:created_at] || Time.now
    @updated_at = attributes[:updated_at] || Time.now
    @attributes = attributes
  end

  def self.find(id)
    storage.find { |record| record.id == id }
  end

  def self.all
    storage.dup
  end

  def self.where(conditions)
    storage.select { |record| conditions.all? { |key, value| record.send(key) == value } }
  end

  def self.create(attributes)
    instance = new(attributes)
    instance.save
    instance
  end

  def save
    touch
    if persisted?
      update_in_storage
    else
      add_to_storage
    end
    self
  end

  def update(attributes)
    @attributes.merge!(attributes)
    attributes.each do |key, value|
      instance_variable_set("@#{key}", value) if respond_to?(key)
    end
    save
  end

  def destroy
    self.class.storage.delete_if { |record| record.id == @id }
    freeze
  end

  def persisted?
    self.class.storage.any? { |record| record.id == @id }
  end

  def to_h
    instance_variables.each_with_object({}) do |var, hash|
      key = var.to_s.delete('@').to_sym
      hash[key] = instance_variable_get(var)
    end
  end

  private

  def generate_id
    SecureRandom.uuid
  end

  def touch
    @updated_at = Time.now
  end

  def self.storage
    @storage ||= []
  end

  def add_to_storage
    self.class.storage << self
  end

  def update_in_storage
    index = self.class.storage.find_index { |record| record.id == @id }
    self.class.storage[index] = self if index
  end
end 

File: fixtures/typescript/test-repo/app/models/base_model.ts
import { randomUUID } from 'crypto';
import { randomUUID as myRandomUUID } from 'crypto';

export class BaseModel {
  public readonly id: string;
  public readonly createdAt: Date;
  public updatedAt: Date;
  protected attributes: Record<string, any>;

  private static storageMap = new Map<string, BaseModel[]>();

  constructor(attributes: Record<string, any> = {}) {
    this.id = attributes.id || randomUUID();
    this.createdAt = attributes.createdAt || new Date();
    this.updatedAt = attributes.updatedAt || new Date();
    this.attributes = attributes;
  }

  static find(id: string): any {
    const storage = this.getStorage();
    return storage.find(record => record.id === id);
  }

  static all(): any[] {
    return [...this.getStorage()];
  }

  static where(conditions: Record<string, any>): any[] {
    const storage = this.getStorage();
    return storage.filter(record => 
      Object.entries(conditions).every(([key, value]) => 
        (record as any)[key] === value
      )
    );
  }

  static create(attributes: Record<string, any>): any {
    const instance = new (this as any)(attributes);
    instance.save();
    return instance;
  }

  save(): this {
    this.touch();
    if (this.persisted()) {
      this.updateInStorage();
    } else {
      this.addToStorage();
    }
    return this;
  }

  update(attributes: Record<string, any>): this {
    Object.assign(this.attributes, attributes);
    Object.entries(attributes).forEach(([key, value]) => {
      if (key in this) {
        (this as any)[key] = value;
      }
    });
    return this.save();
  }

  destroy(): void {
    const storage = (this.constructor as any).getStorage();
    const index = storage.findIndex((record: BaseModel) => record.id === this.id);
    if (index !== -1) {
      storage.splice(index, 1);
    }
    Object.freeze(this);
  }

  persisted(): boolean {
    const storage = (this.constructor as any).getStorage();
    return storage.some((record: BaseModel) => record.id === this.id);
  }

  toObject(): Record<string, any> {
    const result: Record<string, any> = {};
    Object.getOwnPropertyNames(this).forEach(key => {
      if (!key.startsWith('_')) {
        result[key.replace(/^@/, '')] = (this as any)[key];
      }
    });
    return result;
  }

  private touch(): void {
    this.updatedAt = new Date();
  }

  protected static getStorage(): BaseModel[] {
    const className = this.name;
    if (!BaseModel.storageMap.has(className)) {
      BaseModel.storageMap.set(className, []);
    }
    return BaseModel.storageMap.get(className)!;
  }

  private addToStorage(): void {
    const storage = (this.constructor as any).getStorage();
    storage.push(this);
  }

  private updateInStorage(): void {
    const storage = (this.constructor as any).getStorage();
    const index = storage.findIndex((record: BaseModel) => record.id === this.id);
    if (index !== -1) {
      storage[index] = this;
    }
  }
}
